api/ocm/tools/transfer/transfer.go

```diff
@@ -241,113 +243,109 @@ func copyVersion(cctx context.Context, printer common.Printer, log logging.Logge
 	srccd := src.GetDescriptor()
 	cur := *t.GetDescriptor()
 	*t.GetDescriptor() = *prep
-	log.Info("  transferring resources")
-	for i, r := range src.GetResources() {
-		var m ocmcpi.AccessMethod
 
-		if err := common.IsContextCanceled(cctx); err != nil {
-			printer.Printf("cancelled by caller\n")
-			return err
-		}
-
-		nested := finalize.Nested()
-
-		a, err := r.Access()
-		if err == nil {
-			m, err = a.AccessMethod(src)
-			nested.Close(m, fmt.Sprintf("%s: transferring resource %d: closing access method", hist, i))
-		}
-		if err == nil {
-			ok := a.IsLocal(src.GetContext())
-			if !ok {
-				if !none.IsNone(a.GetKind()) {
-					ok, err = handler.TransferResource(src, a, r)
-					if err == nil && !ok {
-						log.Info("transport omitted", "resource", r.Meta().Name, "index", i, "access", a.GetType())
-					}
+	log.Info("  transferring resources and sources using worker pool")
+	tasks := make(chan transferTask)
+	errChan := make(chan error, 100)
+
+	var wg sync.WaitGroup
+	for i := 0; i < maxWorkers; i++ {
+		wg.Add(1)
+		go func(id int) {
+			defer wg.Done()
+			for item := range tasks {
+				if err := item.task(); err != nil {
+					errChan <- err
 				}
 			}
-			if ok {
-				var old compdesc.Resource
-
-				hint := ocmcpi.ArtifactNameHint(a, src)
-				old, err = cur.GetResourceByIdentity(r.Meta().GetIdentity(srccd.Resources))
-
-				changed := err != nil || old.Digest == nil || !old.Digest.Equal(r.Meta().Digest)
-				valueNeeded := err == nil && needsTransport(src.GetContext(), r, &old)
-				if changed || valueNeeded {
-					var msgs []interface{}
-					if !errors.IsErrNotFound(err) {
-						if err != nil {
+		}(i)
+	}
+
+	go func() {
+		for i, r := range src.GetResources() {
+			i, r := i, r
+			tasks <- transferTask{
+				id: fmt.Sprintf("resource-%d", i),
+				task: func() error {
```

Comment:
This functional transfer task is very hard to read and debug, so I suggest you split this into a separate function.